### PR TITLE
[FEAT] Implementation of Resource Management 

### PR DESCRIPTION
**Pull request by CallmeInnocent**
### Description
These API endpoints allows for managing resources within the application. These endpoints will allow for retrieving resources, editing them, creating them and deleting them. 

### How to Test
Access the `GET /api/v1/resources/articles` endpoint to search for published resources returns default page limit for first page
For example:  ` http://localhost:8080/api/v1/resources/articles`
Expected result: Success.
Example response: 

```java
{
    "data": [
        {
            "id": "8cdf7502-1bde-413a-a2b8-7c2490f3a373",
            "title": "Walmart",
            "description": "Sales",
            "image": "hng_images.png",
            "date": "2024-08-14",
            "published": true
        },
        {
            "id": "21594800-85d6-41a6-ba73-32dfac984d17",
            "title": "Amazon",
            "description": "goods",
            "image": "hng_images.png",
            "date": "2024-08-14",
            "published": true
        }
    ],
    "totalPages": 10,
    "currentPage": 0``` 
``` 

Access the ` GET /api/v1/resources/search `  endpoint and encounter a server error.
Expected result: A 500 error is returned.
```java
Example response: {
  "status": "error",
  "message": "Internal Server Error. Please try again later.",
  "status_code": 500
}
``` 

Access the ` POST /api/v1/resources/create `  endpoint to create resources.
For example: `http://localhost:8080/api/v1/resource/create `
Expected result: Created Successfully.
Response:

```java
{
    "message": "How to use youtube  Created Successfully",
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
    }
``` 

Access the ` PATCH  /api/v1/resources/edit `  endpoint to edit already edited resource
For example: `http://localhost:8080/api/v1/resource/edit `
Expected result: Edited Succesfully.
Response:
```java
{
    "message": "How to use youtube  Edited Successfully",
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
    }
``` 

Access the ` DELETE /api/v1/resources/delete/{Id} `  endpoint to delete a resource that doesn’t exist
For example: `http://localhost:8080/api/v1/resource/{id}`
Expected result: Success.

```java
{
    "message": "This resource does not exist",
    "error": "Resource with 3cf9e963-ecd0-42e8-9cd3-c5b0780af8a0 not found",
    "status_code": 404
}
``` 

Access the ` GET  /api/v1/resources/{Id} `  endpoint to get resource with an id
For example: `http://localhost:8080/api/v1/resource/{id}`
Expected result: Success.

```java
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
       }
``` 

Access the ` PATCH  /api/v1/resources/unpublish/{Id} `  endpoint to unpublish a resource
For example: `http://localhost:8080/api/v1/resource/unpublish/{id}`
Expected result: Success.

```java
      "message": "How to use youtube unpublished Successfully"
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": false
         
  }
``` 

Access the ` PATCH  /api/v1/resources/publish/{Id} `  endpoint to publish a resource
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

```java
      "message": "How to use youtube published Successfully"
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
     }
``` 

Access the ` GET  /api/v1/resources/search`  endpoint for an admin to search all resources, published or not.
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "358b9095-6aad-4770-b1c8-b388c361f837",
            "title": "How to make banga soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "1ba89099-449e-44fa-be3c-1cc88c5d6ccf",
            "title": "How to make okra soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        },
        {
            "id": "8f26b20a-9d70-4f6d-aaaa-b32512d702c7",
            "title": "How to make white soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        }
``` 
Access the ` GET  /api/v1/resources/published`  endpoint for an admin to search all published resources
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        }
``` 

Access the ` GET  /api/v1/resources/unpublished`  endpoint for an admin to search all published resources
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        }
```

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    @GetMapping("/searchResources")
+    public ResponseEntity<?> searchResourcesForUsersOnly(
+            @RequestParam() String query,
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "4") int limit) {
+
+        ResourceResponseDto resources = resourceService.findByTitleAndDescriptionForUser(query, PageRequest.of(page, limit));
+        return ResponseEntity.ok(resources);
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    @GetMapping("/articles")
+    public ResponseEntity<?> getAllResourcesForUsersOnly(
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "4") int limit) {
+
+            ResourceResponseDto resources = resourceService.getAllResources(PageRequest.of(page,limit));
+            return ResponseEntity.ok(resources);
+
+    }
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java:29]: Add a space

File: src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        return ResponseEntity.ok(resources);
+
+    }
+
+    @GetMapping("/search")
+    public ResponseEntity<?> searchResourcesForAdminOnly( @RequestParam String query) {
+
+        ResourceResponseDto resources = resourceService.findByTitleAndDescriptionForAdmin(query);
+        return ResponseEntity.ok(resources);
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        ResourceResponseDto addedResources = resourceService.addResources(requestDto);
+        return ResponseEntity.status(HttpStatus.CREATED).body(addedResources);
+    }
+
+    @PatchMapping("/edit")
+    public ResponseEntity<?> editResources(@Valid @RequestBody ResourceRequestDto requestDto){
+
+        ResourceResponseDto editedResources = resourceService.editResources(requestDto);
+        return ResponseEntity.status(HttpStatus.CREATED).body(editedResources);
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/controller/ResourcesController.java:1]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/dto/ResourceResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+@JsonInclude(JsonInclude.Include.NON_NULL)
+
+public class ResourceResponseDto {
+
+    private String message;
+    private List<Resources> data;
+    private Resources resourceData;
+    private Integer totalPages;
+    private Integer currentPage;
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/dto/ResourceResponseDto.java:23]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/entity/Resources.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    @Column(name = "image")
+    private String image;
+
+    @Column(nullable = false, updatable = false)
+    @CreationTimestamp
+    private Date date;
+
+    @Column(name = "published")
+    private Boolean published;
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/entity/Resources.java:38]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/service/ResourceService.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    ResourceResponseDto addResources(ResourceRequestDto resourceRequestDto);
+    ResourceResponseDto deleteResources(String Id);
+    ResourceResponseDto editResources(ResourceRequestDto resourceRequestDto);
+    ResourceResponseDto getResourceById(String Id);
+    ResourceResponseDto unpublishResource(String Id);
+    ResourceResponseDto publishResource(String Id);
+    ResourceResponseDto getAllPublishedResource();
+    ResourceResponseDto findByTitleAndDescriptionForAdmin(String query);
+    ResourceResponseDto getAllUnPublishedResource();
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/service/ResourceService.java:20]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public class ResourceServiceImpl implements ResourceService{
+
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java]: Remove empty space

File: src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        if (resources.isEmpty()) {
+            throw new ResourcesNotFoundException("No resources found with query: " + query);
+        }
+
+        ResourceResponseDto responseDto = new ResourceResponseDto();
+        responseDto.setData(resources);
+
+        return responseDto;
+    }
+    @Override
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java:45]: Add spacing

File: src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+        resourceRepository.save(resources);
+
+        ResourceResponseDto responseDto = new ResourceResponseDto();
+
+        responseDto.setMessage(resources.getTitle() + " Updated successfully");
+        responseDto.setResourceData(resources);
+
+        return responseDto;
+
```
Comment (Am0du) [inline src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java:115]: Remove empty space

File: src/test/java/hng_java_boilerplate/resources/ServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        mockResource.setId(resourceId);
+        mockResource.setPublished(true);
+
+        when(resourceRepository.findById(resourceId)).thenReturn(Optional.of(mockResource));
+        when(resourceRepository.save(any(Resources.class))).thenReturn(mockResource);
+
+
+        ResourceResponseDto response = resourceService.unpublishResource(resourceId);
+
+
```
Comment (Am0du) [inline src/test/java/hng_java_boilerplate/resources/ServiceTest.java]: Remove empty space

File: src/test/java/hng_java_boilerplate/resources/ServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+        String resourceId = "test-resource-id";
+        Resources mockResource = new Resources();
+        mockResource.setId(resourceId);
+        mockResource.setPublished(true);
+
+        when(resourceRepository.findById(resourceId)).thenReturn(Optional.of(mockResource));
+        when(resourceRepository.save(any(Resources.class))).thenReturn(mockResource);
+
+
```
Comment (Am0du) [inline src/test/java/hng_java_boilerplate/resources/ServiceTest.java]: Remove empty space

File: src/test/java/hng_java_boilerplate/resources/ServiceTest.java
Comment (Am0du) [inline src/test/java/hng_java_boilerplate/resources/ServiceTest.java:1]: Remove irrelevant space

**Review by Am0du — CHANGES_REQUESTED** [review]
Remove irrelevant spaces and you're good

File: src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java (hunk 8 above)
Comment (CallmeInnocent) [reply src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java]: from where

File: src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java (hunk 9 above)
Comment (CallmeInnocent) [reply src/main/java/hng_java_boilerplate/resources/service/ResourceServiceImpl.java:45]: spacing to where

**Review by Am0du — APPROVED** [review]

**Review by Am0du — APPROVED** [review]

**Review by Am0du — APPROVED** [review]

**Review by Am0du — APPROVED** [review]

**Pull request merged**